mathics/algorithm/parts.py

```diff
@@ -21,78 +24,84 @@
+        are smaller parts of the corresponding variables at the outer level.
+        """
+        if not indices:
+            return varlist
+
+        if isinstance(varlist, Atom):
+            raise PartDepthError(indices[0])
+
+        pos = indices[0]
+        elements = varlist.elements
+        try:
+            if pos > 0:
+                part = elements[pos - 1]
+            elif pos == 0:
+                part = varlist.get_head()
+            else:
+                part = elements[pos]
+        except IndexError:
+            raise PartRangeError
+        return get_subpart(part, indices[1:])
 
+    return get_subpart(varlist, indices).copy()
 
-def set_part(varlist, indices, newval):
-    "Simple part replacement. indices must be a list of python integers."
 
-    def rec(cur, rest):
-        if len(rest) > 1:
-            pos = rest[0]
-            if isinstance(cur, Atom):
+def set_part(varlist, indices: List[int], newval) -> BaseElement:
+    """Replace the part of ``varlist`` specified by ``indicies`` with the single
+    ``newval`` value. Return the modified compound expression.
+    """
+
+    def set_subpart(varlist, indices: List[int]) -> BaseElement:
+        """Recursive work-horse portion of ``set_part()`` that replaces pieces
+        of ``varlist`` with outer variable ``newval`` based on ``indices``.
+        ``varlist`` and ``indices`` are smaller parts of the corresponding variables at the outer level.
+        """
+        if len(indices) > 1:
+            pos = indices[0]
+            if isinstance(varlist, Atom):
                 raise PartDepthError
             try:
                 if pos > 0:
-                    part = cur.elements[pos - 1]
+                    part = varlist.elements[pos - 1]
                 elif pos == 0:
-                    part = cur.get_head()
+                    part = varlist.get_head()
                 else:
-                    part = cur.elements[pos]
+                    part = varlist.elements[pos]
             except IndexError:
                 raise PartRangeError
-            return rec(part, rest[1:])
-        elif len(rest) == 1:
-            pos = rest[0]
-            if isinstance(cur, Atom):
+            set_subpart(part, indices[1:])
+            return varlist
+        elif len(indices) == 1:
+            pos = indices[0]
+            if isinstance(varlist, Atom):
                 raise PartDepthError
             try:
                 if pos > 0:
-                    cur.set_element(pos - 1, newval)
+                    varlist.set_element(pos - 1, newval)
                 elif pos == 0:
-                    cur.set_head(newval)
+                    if newval == SymbolList:
+                        varlist = ListExpression(varlist._elements)
+                    # If we have other types of specialized compound expressions
+                    # this might be detected as "elif" here.
+                    varlist.set_head(newval)
```

Comment:
`else` ?

Reply:
I don't understand, please elaborate.

Right now, the only customized kind of (compountd) Expression is ListExpression so that is the only place where we need to redo the entire kind of compound Expression. As we add more that can go through this code we'll have to find and adapt those. At some point some more automated way of replacing these without if's will be needed. But we are not there yet. So the comment serves to mark and note such possibility.

Reply:
In a follow-on PR, I see that this code was faulty here: we also need to handle the case for converting from ListExpression to Expression. (Here we convert from Expression to ListExpression only).

The follow-on PR handles creating ListExpression from the outset we go from parser AST to Mathics Expression and this will open up many more opportunities for more streamlined (i.e. faster)  operations. After that I believe we will be able to not have to need `has_form("System`List", None)`, but I believe it opens the possibility of eliminating conversions for list literals in some places.

I am keeping these two PR's separate (at the expense of more work on my side in juggling between versions), so as to keep each PR a little more focused and more intelligible. 

With respect the speedups, I hope there is a lesson getting learned here. We won't get to where we need by low-level tweaks  like changing the order of expressions, or using slots instead of local variables, using Cython, or rewriting the code (where a number of parts are a incomprehensible mess) in C or Rust mirroring the incomprehensible mess.

Instead we'll get there by making this look more like an normal interpreter _first_ and using the long-established practice and techniques there _first_.  Most of the time Slots, homegrown caching, Cython, and stuff like this comes with a lack of readability which was a problem even _before_ any of the low-level speedups which add complexity and confusion were added. 

To me this is doing a final painting the outside or interior of a house before you've added the roof or finished constructing most of the walls interior. 

_After_ we get this to look like a normal interpreter, all of those things can be done. Rewriting in Rust might give a 10 times improvement, I wouldn't undertake doing something like the in advance of cleaning up the mess that is here now.

Reply:
```
if newval == SymbolList:
     varlist = ListExpression(varlist._elements)
# else:
     varlist.set_head(newval)
```

You shouldn't modify the head of ListExpression

Reply:
That's true. 89e97ea5 adds some checks for this.